### ISSUE TITLE
gpu-operator injecting runtimeClass after transitioning containerd runtime node

### ISSUE THREAD

**Issue by dcarrion87** [NVIDIA/gpu-operator#477]
We're currently running k3s and wanting to transition from docker to containerd runtime.

No issues except for the gpu-operator pods which insist on injecting runtimeClass into it's pods and causing:

```
  Warning  FailedCreatePodSandBox  5s (x7 over 83s)  kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = RuntimeHandler "nvidia" not supported
```

Which breaks the operator pods on all the existing docker nodes as soon as I add a transitioned containerd node into the cluster.

How do I stop it from doing this during the transition?

Thanks

**Comment by klueska** [comment]
I would argue that the operator should probably unconditionally deploy the runtime class regardless of the runtime it "thinks" is in use in the cluster, but one way out of this situation would be to manually create the runtimeclass yourself:
```
apiVersion: node.k8s.io/v1
handler: /usr/local/nvidia/toolkit/nvidia-container-runtime
kind: RuntimeClass
metadata:
  labels:
    app.kubernetes.io/component: gpu-operator
  name: nvidia
```

**Comment by dcarrion87** [comment]
Thanks @klueska. 

The runtimeclass already exists with and without that node in there. E.g.:

```
apiVersion: v1
items:
- apiVersion: node.k8s.io/v1
  handler: nvidia
  kind: RuntimeClass
  metadata:
    creationTimestamp: "2022-11-23T07:22:41Z"
    labels:
      app.kubernetes.io/component: gpu-operator
    name: nvidia
    ownerReferences:
    - apiVersion: nvidia.com/v1
      blockOwnerDeletion: true
      controller: true
      kind: ClusterPolicy
      name: cluster-policy
      uid: d7c409fc-fba3-4d5d-8d74-3e716f7ef3ef
    resourceVersion: "99192"
    uid: aaf727bb-9ffc-4548-ae3a-de9603781fe4
kind: List
metadata:
  resourceVersion: ""
```

The issue is that only after we added the containerd node it started injecting "runtimeClass: nvidia" into it's pods which is causing trouble on our docker nodes.

I'm just totally confused why it would do this only after we tried to transition a node.

**Comment by klueska** [comment]
Actually, re-reading your issue -- it seems the RuntimeClass is installed, but your containerd config doesn't seem to have a reference to a runtime called `nvidia`.

Did you deploy the operator with the following settings so that it knows which config to modify to add the nvidia runtime? On k3s this must be specified because it's not the default location for containerd.

```
    toolkit:
      env:
      - name: CONTAINERD_CONFIG
        value: /var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl
      - name: CONTAINERD_SOCKET
        value: /run/k3s/containerd/containerd.sock
      - name: CONTAINERD_RUNTIME_CLASS
        value: nvidia
      - name: CONTAINERD_SET_AS_DEFAULT
        value: "true"
```

**Comment by dcarrion87** [comment]
E.g. Before adding node:

```
➜  ~ kubectl get pod -n gpu-operator nvidia-dcgm-exporter-sc2mm -o yaml | grep runtimeClass
```

After adding node:

```
➜  ~ kubectl get pod -n gpu-operator nvidia-dcgm-exporter-dvlm6 -o yaml | grep runtimeClass
  runtimeClassName: nvidia
```

I don't want it configuring anything or injecting anything. We configure via ansible.

**Comment by klueska** [comment]
OK, we'd need to check with @shivamerla and @cdesiniotis as to what's going on here.

I know that (in theory) we are supposed to support a mix of runtimes in the cluster, but it seems that as soon as a containerd node gets added, it assumes that all "management" pods can be run with a reference to the `nvidia` runtimeclass (which is not true if docker is the runtime in use as it does not support the notion of a runtimeclass).

**Comment by dcarrion87** [comment]
Thanks @klueska . I appreciate that it's not ideal to have to support this. 

I'm taking a peak around the code to see where this logic happens / where it decides to now add runtimeClassName to see if there's a work around to not add it in by default.

**Comment by dcarrion87** [comment]
@klueska 

I'm guessing this might be of interest?

containerd if >=1 node: https://github.com/NVIDIA/gpu-operator/blob/a438e0060666136e94ee5a2c244be3078a5c4639/controllers/state_manager.go#L700


I can see transformations here but not sure I understand exactly how it related: https://github.com/NVIDIA/gpu-operator/blob/4268fc5fb7f435dac2842c7fe501f07a74590dd0/controllers/object_controls.go

**Comment by dcarrion87** [comment]
Here's my thoughts on a potential workaround. Your thoughts welcomed:


Add a handler called docker to containerd config:
```
[plugins.opt]
  path = "/var/lib/rancher/k3s/agent/containerd"
....
  [plugins.cri.containerd.runtimes."docker"]
    runtime_type = "io.containerd.runc.v2"
    [plugins.cri.containerd.runtimes."docker".options]
      SystemdCgroup = true
      BinaryName = "/usr/bin/nvidia-container-runtime"
```

Specify handler as docker in thee nvidia runtimeclass the gpu-operator creates

- When run on containerd node it will just work
- When run on docker node it will fallback to legacy handling

```
apiVersion: v1
items:
- apiVersion: node.k8s.io/v1
  handler: docker
  kind: RuntimeClass
  metadata:
    labels:
      app.kubernetes.io/component: gpu-operator
    name: nvidia
kind: List
```

When transition is over:

- Pull the docker containerd runtime handler out of the config
- Delete the runtimeclass and let the gpu-operator controller re-create.

**Comment by dcarrion87** [comment]
@klueska @shivamerla @cdesiniotis we implemented the workaround discussed here and it seems to work:

https://github.com/NVIDIA/gpu-operator/issues/477#issuecomment-1401570941

Happy for this to close unless you see it worth making the operator more tolerable of multiple runtimes.

**Comment by cdesiniotis** [comment]
@dcarrion87 your workaround seems logical to me. Thank you for providing the details and confirming it works.

Unfortunately we do not support different container runtimes in the cluster which is the crux of this issue. The toolkit container depends on the `RUNTIME` envvar to be set which determines the container runtime it attempts to configure. We set this envvar in the operator here: https://github.com/NVIDIA/gpu-operator/blob/4268fc5fb7f435dac2842c7fe501f07a74590dd0/controllers/object_controls.go#L1000-1003

Ideally we enhance the toolkit container to detect the container runtime used on the node and configure it accordingly. This still doesn't address the RuntimeClass issue, which needs some more thought.

**Comment by cdesiniotis** [comment]
In the interest of cleaning up our issue backlog, I am closing this stale issue as suggested in https://github.com/NVIDIA/gpu-operator/issues/477#issuecomment-1403018736. We don't have any plans to make the operator more tolerable to different types of container runtimes in the cluster. But feel free to open a new issue if you still want to request this feature.
